File: core/src/com/sergey/spacegame/client/ecs/component/VisualComponent.java
package com.sergey.spacegame.client.ecs.component;

import com.badlogic.ashley.core.Component;
import com.badlogic.ashley.core.ComponentMapper;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;
import com.sergey.spacegame.SpaceGame;
import com.sergey.spacegame.common.ecs.component.ClonableComponent;

import java.lang.reflect.Type;

public class VisualComponent implements ClonableComponent {
	public static final ComponentMapper<VisualComponent> MAPPER = ComponentMapper.getFor(VisualComponent.class);
	
	private TextureRegion region;
	private String name;
	
	public VisualComponent() {}

	public VisualComponent(String name) {
		setRegion(name);
	}

	@Override
	public Component copy() {
		return new VisualComponent(name);
	}
	
	public TextureRegion getRegion() {
		return region;
	}
	
	public void setRegion(String name) {
		this.region = SpaceGame.getInstance().getRegion(name);
		this.name = name;
	}
	
	public String getName() {
		return name;
	}
	
	public static class Adapter implements JsonSerializer<VisualComponent>, JsonDeserializer<VisualComponent> {

		@Override
		public VisualComponent deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
			JsonObject obj = json.getAsJsonObject();
			
			return new VisualComponent(obj.get("image").getAsString());
		}

		@Override
		public JsonElement serialize(VisualComponent src, Type typeOfSrc, JsonSerializationContext context) {
			JsonObject obj = new JsonObject();
			
			obj.addProperty("image", src.name);

			return obj;
		}
	}
}

File: core/src/com/sergey/spacegame/common/ecs/component/VisualComponent.kt
package com.sergey.spacegame.common.ecs.component

import com.badlogic.ashley.core.Component
import com.badlogic.ashley.core.ComponentMapper
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import com.sergey.spacegame.common.SpaceGame
import com.sergey.spacegame.common.data.VisualData
import java.lang.reflect.Type

/**
 * This component contains the visual data
 *
 * @author sergeys
 *
 * @constructor Create a new VisualComponent object
 *
 * @param regionName - the name of the texture region
 */
class VisualComponent(regionName: String) : ClonableComponent {
    
    /**
     * The texture region's name
     */
    var regionName: String = regionName
        set(value) {
            visualData = SpaceGame.getInstance().context.createVisualData(value)
            field = value
        }
    
    /**
     * The visual data object used for client side rendering
     */
    var visualData: VisualData? = SpaceGame.getInstance().context.createVisualData(regionName)
        private set
    
    override fun copy(): Component {
        return VisualComponent(regionName)
    }
    
    companion object MAP {
        @JvmField
        val MAPPER = ComponentMapper.getFor(VisualComponent::class.java)!!
    }
    
    class Adapter : JsonSerializer<VisualComponent>, JsonDeserializer<VisualComponent> {
        
        @Throws(JsonParseException::class)
        override fun deserialize(json: JsonElement, typeOfT: Type,
                                 context: JsonDeserializationContext): VisualComponent {
            val obj = json.asJsonObject
            
            return VisualComponent(obj.get("image").asString)
        }
        
        override fun serialize(src: VisualComponent, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
            val obj = JsonObject()
            
            obj.addProperty("image", src.regionName)
            
            return obj
        }
    }
}